cpu/sam3x: extend gpio implementation

FrancescoErmini opened the pull request:
-Extended GPIO from 16 to 31 for Arduino DUE/UDOO board, with cpu sam3x8e
-add two missing functions in gpio.c: gpio_irq_disable(gpio_t dev) and gpio_irq_enable(gpio_t dev)

NOTE: updated exel file in Arduino Due Wiki https://docs.google.com/spreadsheets/d/11Q33vhu9o_jv40dnQS9exVzZlCcX5oyiVkSpxZkExH4/edit?usp=sharing

bapclenet commented:
@OlegHahm, I don't have an Arduino DUE/UDOO board so I'm not reviewing this PR, I mostly focus on the SAMR21. How to be anassigned?

PeterKietzmann commented:
Okee, will put it on my ToDo list

OlegHahm commented:
@bapclenet, sorry, I didn't look carefully.
@PeterKietzmann, thanks!

PeterKietzmann commented:
Tested, works. Please squash your commits. Maybe I'll do a short PR to extend the test in tests/periph_gpio_exti. In your table the the pins `GPIO_22 ... GPIO_25` are missing. I'll ask @haukepetersen if he could give you write access to adapt [this table](https://docs.google.com/spreadsheets/d/1g5xMpLbi6cMc9QdHQrpf3XH_hiRb4wrAu1wvvpelMUw/edit) or if he could add your changes

FrancescoErmini commented:
Hi Peter,

GPIO_22 - GIPO_25 were together with PWML0 - PWML3, and didn't have a
yellow background for that reason.
I've modify the exel file, I add an other column 'function2' for pins that
have double functionality. I hope now it's more clear.
@haukepetersen <https://github.com/haukepetersen>, can modify, download and
upload the exel file that I've shared in my Google Drive without problems.
I also modify the Arduino Due image in the Wiki:
https://drive.google.com/file/d/0B0XninFV9C0bU2hOLXIwbmRsS0E/view?usp=sharing





2015-04-14 11:32 GMT+02:00 Peter Kietzmann <notifications@github.com>:

> Tested, works. Please squash your commits. Maybe I'll do a short PR to
> extend the test in tests/periph_gpio_exti. In your table the the pins GPIO_22
> ... GPIO_25 are missing. I'll ask @haukepetersen
> <https://github.com/haukepetersen> if he could give you write access to
> adapt this table
> <https://docs.google.com/spreadsheets/d/1g5xMpLbi6cMc9QdHQrpf3XH_hiRb4wrAu1wvvpelMUw/edit>
> or if he could add your changes
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/2798#issuecomment-92717716>.
>



-- 
Francesco Ermini
Via Abebe Bikila, 8 50012 (FI)
cell. 3338710609
tell. 055642820

PeterKietzmann commented:
@FrancescoErmini thanks for the schematic and the table, I'll update our wiki when this PR is merged. According to you pin-configuration: in general we try to avoid having pins that can act in different roles, to prevent conflicts. Do you really need 32 pins?  Otherwise we could think of moving `PWM0...PWM3`. This needs to be discussed in short, but the benefit is that we gert a board with a complete GPIO header!

PeterKietzmann commented:
Aaah and could you trim your commit message a bit please? You could use something like: "cpu/sam3x: extend gpio implementation"

FrancescoErmini commented:
no I don't need 32 GPIOs, I only needed some GPIOs on the 'internal side' of the UDOO board , because  I'm making a shield that requires those connections. I shared the code I wrote for my self with a pull request, but I agree to have different roles in different pins. I think that in most cases 16 GPIOs are sufficient, and I don't want to force the use of 32 GPIOs.  So tell me what's better for the RIOT project; write GPIO [16-21] [26-31] (excluding PWM0..3) or even let the pull request fails to keep the RIOT philosophy of 15 GPIO. If a user need some extra pins, then It's very easy to write its code.

PeterKietzmann commented:
No, disregarding this PR isn't the way to go and would be useless work. Also I like the idea of having a whole connector with GPIOs. I'd vote for moving the affected PWM stuff to other pins (I think there are some pins available on the right side of the board). But therefore I need to check (i) if this is technically possible and (ii) if there are objections from other developers and users. So let's keep this PR and I'll trigger the next steps when there is any progress. Ok?

FrancescoErmini commented:
Ok I'll wait new instructions.
For now I changed the commit as you told me and removed the second column in the exel file.
PWMs on the right side are located at pin 6 - 9 but I noticed that in cpu/sam3x8e/periph/  the file pwm.c is missing...and the exel sheet rows of PWMs quote  'not connected'.

PeterKietzmann commented:
Ah I didn't look at the source code. So that could simplify the situation. The pwm seems to be not yet implemented. Still I'd like to hear another opinion on this. @haukepetersen do you agree to use all pins of the "double-row" at the bottom of the arduino-due for GPIOs? There is nothing else implemented on these pins. And as far as I see it the only peripheral that COULD be affected is the PWM which can also be implemented on some other pins on the right side of the board.

haukepetersen commented:
This PR is fine - using all the pins on the bottom of the board for GPIO is ok with me. Though I argue that in no real application one really needs >16 GPIOs :-)

Anyway: ACK and go.
